spec/Section 6 -- Execution.md

```diff
@@ -42,44 +73,106 @@ of variable's declared type. If a query error is encountered during
+  * Let {queryType} be the root Query type in {schema}.
+  * Assert: {queryType} is an Object type.
+  * Let {selectionSet} be the top level Selection Set in {query}.
+  * Let {data} be the result of running
+    {ExecuteSelectionSet(selectionSet, queryType, initialValue, variableValues)}
+    *normally* (allowing parallelization).
+  * Let {errors} be any *field errors* produced while executing the
+    selection set.
+  * Return an unordered map containing {data} and {errors}.
+
 If the operation is a mutation, the result of the operation is the result of
-evaluating the mutation’s top level selection set on the “Mutation Root”
-object. This selection set should be evaluated serially.
+executing the mutation’s top level selection set on the “Mutation Root”
+object. This selection set should be executed serially.
 
-If the operation is a query, the result of the operation is the result of
-evaluating the query’s top level selection set on the “Query Root” object.
+It is expected that the top level fields in a mutation operation perform
+side-effects on the underlying data system. Serial execution of the provided
+mutations ensures against race conditions during these side-effects.
+
+ExecuteMutation(mutation, schema, variableValues, initialValue):
+
+  * Let {variableValues} be the set of variable values to be used by any
+    field argument value coercion.
+  * Let {mutationType} be the root Mutation type in {schema}.
+  * Assert: {mutationType} is an Object type.
+  * Let {selectionSet} be the top level Selection Set in {mutation}.
+  * Let {data} be the result of running
+    {ExecuteSelectionSet(selectionSet, mutationType, initialValue, variableValues)}
+    *serially*.
+  * Let {errors} be any *field errors* produced while executing the
+    selection set.
+  * Return an unordered map containing {data} and {errors}.
+
+
+## Executing Selection Sets
 
+To execute a selection set, the object value being evaluated and the object type
+need to be known, as well as whether it must be executed serially, or may be
+executed in parallel.
 
-## Evaluating selection sets
+First, the selection set is turned into a grouped field set; then, each
+represented field in the grouped field set produces an entry into a
+response map.
 
-To evaluate a selection set, the executor needs to know the object on which it
-is evaluating the set and whether it is being evaluated serially.
+ExecuteSelectionSet(selectionSet, objectType, objectValue, variableValues):
 
-If the selection set is being evaluated on the `null` object, then the result
-of evaluating the selection set is `null`.
+  * Initialize {visitedFragments} to be the empty set.
+  * Let {groupedFieldSet} be the result of
+    {CollectFields(objectType, selectionSet, visitedFragments, variableValues)}.
+  * Initialize {resultMap} to an empty ordered map.
+  * For each {groupedFields} in {groupedFieldSet}:
+    * Let {entryTuple} be {GetFieldEntry(objectType, objectValue, groupedFields, variableValues)}.
```

Comment:
The tuple approach was decided when @dschafer and I were writing this section last year. Namely we need to distinguish between a field value of null vs a field not being included at all. Otherwise as you're pointing out, a larger part of the `GetFieldEntry` would end up in this algorithm instead. It aligns to the reference implementation (which in JS returns either `null` as a value or `undefined` as a skip).

I like this suggestion though - I'd like to explore this as a follow up since it might also suggest changing the reference implementation as well.